component/stream/read_write_filename.go

```diff
@@ -0,0 +1,478 @@
+		return err
+	}
+	if !rw.StreamOnly {
+		rw.purge(options.Handle.Path, true)
+	}
+	err = rw.NextComponent().CloseFile(options)
+	if err != nil {
+		log.Err("Stream::CloseFile : error closing file %s [%s]", options.Handle.Path, err.Error())
+	}
+	return err
+}
+
+func (rw *ReadWriteFilenameCache) FlushFile(options internal.FlushFileOptions) error {
+	// log.Trace("Stream::CloseFile : name=%s, handle=%d", options.Handle.Path, options.Handle.ID)
+	err := rw.NextComponent().FlushFile(options)
+	if err != nil {
+		log.Err("Stream::FlushFile : error flushing file %s [%s]", options.Handle.Path, err.Error())
+		return err
+	}
+	options.Handle.Flags.Clear(handlemap.HandleFlagDirty)
+	return nil
+}
+
+// TODO: any cached file to be purged without flushing
+func (rw *ReadWriteFilenameCache) DeleteFile(options internal.DeleteFileOptions) error {
+	// log.Trace("Stream::DeleteFile : name=%s", options.Name)
+	if !rw.StreamOnly {
+		rw.purge(options.Name, false)
+	}
+	err := rw.NextComponent().DeleteFile(options)
+	if err != nil {
+		log.Err("Stream::DeleteFile : error deleting file %s [%s]", options.Name, err.Error())
+		return err
+	}
+	return nil
+}
+
+// TODO: any cached file to be purged without flushing
+func (rw *ReadWriteFilenameCache) DeleteDirectory(options internal.DeleteDirOptions) error {
+	log.Trace("Stream::DeleteDirectory : name=%s", options.Name)
+	for fileName := range rw.fileCache {
+		if strings.HasPrefix(fileName, options.Name) {
+			rw.purge(fileName, false)
+		}
+	}
+	err := rw.NextComponent().DeleteDir(options)
+	if err != nil {
+		log.Err("Stream::DeleteDirectory : error deleting directory %s [%s]", options.Name, err.Error())
+		return err
+	}
+	return nil
+}
+
+// TODO: any cached file to be purged without flushing
+func (rw *ReadWriteFilenameCache) RenameDirectory(options internal.RenameDirOptions) error {
+	log.Trace("Stream::RenameDirectory : name=%s", options.Src)
+	for fileName := range rw.fileCache {
+		if strings.HasPrefix(fileName, options.Src) {
+			rw.purge(fileName, false)
+		}
+	}
+	err := rw.NextComponent().RenameDir(options)
+	if err != nil {
+		log.Err("Stream::RenameDirectory : error renaming directory %s [%s]", options.Src, err.Error())
+		return err
+	}
+	return nil
+}
+
+// Stop : Stop the component functionality and kill all threads started
+func (rw *ReadWriteFilenameCache) Stop() error {
+	log.Trace("Stopping component : %s", rw.Name())
+	if !rw.StreamOnly {
+		rw.Lock()
+		defer rw.Unlock()
+		for fileName, buffer := range rw.fileCache {
+			delete(rw.fileCache, fileName)
+			buffer.Lock()
+			defer buffer.Unlock()
+			buffer.Purge()
+			atomic.AddInt32(&rw.CachedObjects, -1)
+		}
+	}
+	return nil
+}
+
+// GetAttr : Try to serve the request from the attribute cache, otherwise cache attributes of the path returned by next component
+func (rw *ReadWriteFilenameCache) GetAttr(options internal.GetAttrOptions) (*internal.ObjAttr, error) {
+	// log.Trace("AttrCache::GetAttr : %s", options.Name)
```

Comment:
We need to handle the dir listing operation as well? there might be case where a new file is created but its not uploaded to container and in such case list operation on dir shall be showing those files as well. If this is a miss, we shall recheck rw-streaming layer also for the same miss.